### PR TITLE
Update go version to 1.20.5

### PR DESCRIPTION
**Pull request by Kaderinho**
### What does this PR do?

Update the Go version used for Agent to 1.20.5

### Motivation

Using supported minor version

### Describe how to test/QA your changes

No specific tests but important QA scrutiny to avoid introducing potential regressions (RSS, cpu/threading, GC behavior, etc)

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: .gitlab-ci.yml [hunk 1]
```diff
@@ -146,19 +146,19 @@ variables:
   # To use images from datadog-agent-buildimages dev branches, set the corresponding
   # SUFFIX variable to _test_only
   DATADOG_AGENT_BUILDIMAGES_SUFFIX: ""
-  DATADOG_AGENT_BUILDIMAGES: v16742019-a2675dc
+  DATADOG_AGENT_BUILDIMAGES: v17115340-e510775
```
Comment (KSerrania) [inline .gitlab-ci.yml:149]: Note: to be updated before merging (this isn't the latest `datadog-agent-buildimages` commit anymore).

File: .circleci/config.yml [hunk 2]
```diff
@@ -257,6 +257,7 @@ jobs:
 
   build_binaries:
     <<: *job_template
+    resource_class: xlarge
```
Comment (KSerrania) [inline .circleci/config.yml:260]:
```suggestion
    resource_class: large
```
should be enough?

File: tasks/test.py [hunk 3]
```diff
@@ -302,6 +302,32 @@ def command(module_results, module, module_result):
...
+    args: Dict[str, str],
+    test_profiler: TestProfiler,
+):
+    """
+    Builds the stdlib with the same build flags as the tests.
+    """
+    args["go_build_tags"] = " ".join(build_tags)
+
+    print("GOCACHE size before")
+    ctx.run("du -sch $(go env GOCACHE)", warn=True)
```
Comment (KSerrania) [inline tasks/test.py:319]: These were debug logs, they can be removed.

File: tasks/test.py [hunk 4]
```diff
@@ -302,6 +302,32 @@ def command(module_results, module, module_result):
...
+    ctx.run(
+        cmd.format(
+            **args
+        ),
+        env=env,
+        out_stream=test_profiler,
+        warn=True,
+    )
+    print("GOCACHE size after")
+    ctx.run("du -sch $(go env GOCACHE)", warn=True)
```
Comment (KSerrania) [inline tasks/test.py:329]: These were debug logs, they can be removed.

File: tasks/test.py [hunk 5]
```diff
@@ -302,6 +302,32 @@ def command(module_results, module, module_result):
...
+def build_stdlib(
+    ctx,
+    build_tags: List[str],
+    cmd: str,
+    env: Dict[str, str],
+    args: Dict[str, str],
+    test_profiler: TestProfiler,
+):
+    """
+    Builds the stdlib with the same build flags as the tests.
```
Comment (KSerrania) [inline tasks/test.py:314]: Could you expand the description a bit to explain why we need to do this?

Comment (Kaderinho) [reply tasks/test.py:314]: Added some context in https://github.com/DataDog/datadog-agent/pull/17744/commits/dd78008e32da94f86f633a69775cb801b483d200. Please let me know if this is good 🙏

**Review by vickenty — APPROVED** [review]

**Review by kayayarai — APPROVED** [review]

**Review by hush-hush — APPROVED** [review]

**Review by pgimalac — APPROVED** [review]

**Review by katiehockman — APPROVED** [review]

**Review by amenasria — APPROVED** [review]
LGTM !!

**Pull request merged**